companion.py

```diff
@@ -463,30 +528,127 @@ def base64_url_encode(self, text: bytes) -> str:
         return base64.urlsafe_b64encode(text).decode().replace('=', '')
 
 
+class EDMCCAPIReturn:
```

Comment:
This could be a dataclass, or namedtuple

Reply:
I don't think it could be a namedtuple, as it's the base class for the Request, Response and Failed classes.

I'll check if they can all be dataclasses though.  Looks like I can lose the `super().__init__()` calls if so.

Reply:
Nope, due to the mix of defaulted and non-defaulted variables that's proving problematic with inheritance:

```python
@dataclass
class Base:

  var1: int
  var2: bool = False


class Inherits(Base):

  var3: str
```
Leads to it complaining that `Inherits.var3` "follows default argument", because it's effectively after `var2`.